File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/io/Io.java
package app.zxtune.io;

import android.os.Build;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.RandomAccessFile;
import java.nio.BufferOverflowException;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;

import app.zxtune.Log;

public class Io {
  private static final String TAG = Io.class.getName();

  public static final int MIN_MMAPED_FILE_SIZE = 131072;
  public static final int INITIAL_BUFFER_SIZE = 262144;

  public static ByteBuffer readFrom(File file) throws IOException {
    final FileInputStream stream = new FileInputStream(file);
    try {
      final FileChannel channel = stream.getChannel();
      try {
        return readFrom(channel);
      } finally {
        channel.close();
      }
    } finally {
      stream.close();
    }
  }

  private static ByteBuffer readFrom(FileChannel channel) throws IOException {
    if (channel.size() >= MIN_MMAPED_FILE_SIZE) {
      for (int retry = 1; ; ++retry) {
        try {
          return readMemoryMapped(channel);
        } catch (IOException e) {
          if (retry == 1) {
            Log.w(TAG, e, "Failed to read using MMAP. Cleanup memory");
            //http://stackoverflow.com/questions/8553158/prevent-outofmemory-when-using-java-nio-mappedbytebuffer
            System.gc();
            System.runFinalization();
          } else {
            Log.w(TAG, e, "Failed to read using MMAP. Fallback");
            break;
          }
        }
      }
    }
    return readDirectArray(channel);
  }

  private static ByteBuffer readMemoryMapped(FileChannel channel) throws IOException {
    return channel.map(FileChannel.MapMode.READ_ONLY, 0, validateSize(channel.size()));
  }

  private static ByteBuffer readDirectArray(FileChannel channel) throws IOException {
    final ByteBuffer direct = allocateDirectBuffer(channel.size());
    channel.read(direct);
    direct.position(0);
    return direct;
  }

  private static long validateSize(long size) throws IOException {
    if (size == 0) {
      throw new IOException("Empty file");
    } else {
      return size;
    }
  }

  public static ByteBuffer readFrom(InputStream stream) throws IOException {
    try {
      byte[] buffer = reallocate(null);
      int size = 0;
      for (; ; ) {
        size = readPartialContent(stream, buffer, size);
        if (size == buffer.length) {
          buffer = reallocate(buffer);
        } else {
          break;
        }
      }
      validateSize(size);
      return ByteBuffer.wrap(buffer, 0, size);
    } finally {
      stream.close();
    }
  }

  public static ByteBuffer readFrom(InputStream stream, long size) throws IOException {
    try {
      final ByteBuffer result = allocateDirectBuffer(size);
      final byte[] buffer = reallocate(null);
      int totalSize = 0;
      for (; ; ) {
        final int partSize = readPartialContent(stream, buffer, 0);
        if (partSize != 0) {
          totalSize += partSize;
          result.put(buffer, 0, partSize);
          if (partSize == buffer.length) {
            continue;
          }
        }
        break;
      }
      if (totalSize != size) {
        throw new IOException("File size mismatch");
      }
      result.position(0);
      return result;
    } catch (BufferOverflowException err) {
      throw new IOException("File size mismatch", err);
    } finally {
      stream.close();
    }
  }

  private static int readPartialContent(InputStream stream, byte[] buffer, int offset) throws IOException {
    final int len = buffer.length;
    while (offset < len) {
      final int chunk = stream.read(buffer, offset, len - offset);
      if (chunk < 0) {
        break;
      }
      offset += chunk;
    }
    return offset;
  }

  private static byte[] reallocate(@Nullable byte[] buf) throws IOException {
    for (int retry = 1; ; ++retry) {
      try {
        if (buf != null) {
          final byte[] result = new byte[buf.length * 3 / 2];
          System.arraycopy(buf, 0, result, 0, buf.length);
          return result;
        } else {
          return new byte[INITIAL_BUFFER_SIZE];
        }
      } catch (OutOfMemoryError err) {
        if (retry == 1) {
          Log.d(TAG, "Retry reallocate call for OOM");
          System.gc();
        } else {
          throw new IOException(err);
        }
      }
    }
  }

  private static ByteBuffer allocateDirectBuffer(long size) throws IOException {
    validateSize(size);
    for (int retry = 1; ; ++retry) {
      try {
        return ByteBuffer.allocateDirect((int) size);
      } catch (OutOfMemoryError err) {
        if (retry == 1) {
          Log.d(TAG, "Retry reallocate call for OOM");
          System.gc();
        } else {
          throw new IOException(err);
        }
      }
    }
  }

  public static long copy(InputStream in, OutputStream out) throws IOException {
    try {
      final byte[] buffer = reallocate(null);
      for (long total = 0; ; ) {
        final int size = readPartialContent(in, buffer, 0);
        out.write(buffer, 0, size);
        total += size;
        if (size != buffer.length) {
          return total;
        }
      }
    } finally {
      in.close();
    }
  }

  public static void touch(File file) {
    if (!file.setLastModified(System.currentTimeMillis())) {
      try {
        final RandomAccessFile raf = new RandomAccessFile(file, "rw");
        final long size = raf.length();
        raf.setLength(size + 1);
        raf.setLength(size);
        raf.close();
      } catch (IOException e) {
        Log.w(TAG, e, "Failed to update file timestamp");
      }
    }
  }

  static boolean rename(File oldName, File newName) {
    if (Build.VERSION.SDK_INT >= 26) {
      return renameViaFiles(oldName, newName);
    } else {
      return oldName.renameTo(newName)
          || (newName.exists() && newName.delete() && oldName.renameTo(newName));
    }
  }

  @RequiresApi(26)
  private static boolean renameViaFiles(File oldName, File newName) {
    try {
      Files.move(oldName.toPath(), newName.toPath(), StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
      return true;
    } catch (IOException e) {
      Log.w(TAG, e, "Failed to rename '%s' -> '%s'", oldName.getAbsolutePath(), newName.getAbsolutePath());
    }
    return false;
  }

  public static InputStream createByteBufferInputStream(final ByteBuffer buf) {
    return new InputStream() {

      @Override
      public int available()  {
        return buf.remaining();
      }

      @Override
      public void mark(int readLimit) {
        buf.mark();
      }

      @Override
      public void reset() {
        buf.reset();
      }

      @Override
      public boolean markSupported() {
        return true;
      }

      @Override
      public int read() {
        return buf.hasRemaining()
          ? buf.get()
          : -1;
      }

      @Override
      public int read(byte[] bytes, int off, int len) {
        if (buf.hasRemaining()) {
          len = Math.min(len, buf.remaining());
          buf.get(bytes, off, len);
          return len;
        } else {
          return -1;
        }
      }
    };
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/io/Io.kt
package app.zxtune.io

import android.os.Build
import androidx.annotation.RequiresApi
import app.zxtune.Logger
import java.io.*
import java.nio.ByteBuffer
import java.nio.channels.FileChannel
import java.nio.file.Files
import java.nio.file.StandardCopyOption

private val LOG = Logger(Io::class.java.name)

object Io {

    const val MIN_MMAPED_FILE_SIZE = 131072
    const val INITIAL_BUFFER_SIZE = 262144

    @JvmStatic
    @Throws(IOException::class)
    fun readFrom(file: File) = readFrom(FileInputStream(file))

    @JvmStatic
    @Throws(IOException::class)
    fun readFrom(stream: FileInputStream) = stream.use {
        stream.channel.use {
            readFrom(it)
        }
    }

    private fun readFrom(channel: FileChannel): ByteBuffer {
        if (channel.size() >= MIN_MMAPED_FILE_SIZE) {
            var retry = 1
            while (true) {
                try {
                    return readMemoryMapped(channel)
                } catch (e: IOException) {
                    if (retry == 1) {
                        LOG.w(e) { "Failed to read using MMAP. Cleanup memory" }
                        //http://stackoverflow.com/questions/8553158/prevent-outofmemory-when-using-java-nio-mappedbytebuffer
                        System.gc()
                        System.runFinalization()
                    } else {
                        LOG.w(e) { "Failed to read using MMAP. Fallback" }
                        break
                    }
                }
                ++retry
            }
        }
        return readDirectArray(channel)
    }

    private fun readMemoryMapped(channel: FileChannel) =
        channel.map(FileChannel.MapMode.READ_ONLY, 0, validateSize(channel.size()))

    private fun readDirectArray(channel: FileChannel) = allocateDirectBuffer(channel.size()).apply {
        channel.read(this)
        position(0)
    }

    private fun validateSize(size: Long) = if (size != 0L) size else throw IOException("Empty file")

    @JvmStatic
    @Throws(IOException::class)
    fun readFrom(stream: InputStream): ByteBuffer = stream.use {
        var buffer = reallocate(null)
        var size = 0
        while (true) {
            size = readPartialContent(stream, buffer, size)
            buffer = if (size == buffer.size) {
                reallocate(buffer)
            } else {
                break
            }
        }
        validateSize(size.toLong())
        return ByteBuffer.wrap(buffer, 0, size)
    }

    @JvmStatic
    fun readFrom(stream: InputStream, size: Long): ByteBuffer = stream.use {
        val result = allocateDirectBuffer(size)
        val buffer = reallocate(null)
        var totalSize = 0L
        while (true) {
            val partSize = readPartialContent(stream, buffer, 0)
            if (partSize != 0) {
                totalSize += partSize
                if (totalSize > size) {
                    throw IOException("File size mismatch")
                }
                result.put(buffer, 0, partSize)
                if (partSize == buffer.size) {
                    continue
                }
            }
            break
        }
        if (totalSize != size) {
            throw IOException("File size mismatch")
        }
        result.position(0)
        return result
    }

    private fun readPartialContent(stream: InputStream, buffer: ByteArray, offset: Int): Int {
        var pos = offset
        val len = buffer.size
        while (pos < len) {
            val chunk = stream.read(buffer, pos, len - pos)
            if (chunk < 0) {
                break
            }
            pos += chunk
        }
        return pos
    }

    private fun reallocate(buf: ByteArray?): ByteArray {
        var retry = 1
        while (true) {
            try {
                return buf?.copyOf(buf.size * 3 / 2) ?: ByteArray(INITIAL_BUFFER_SIZE)
            } catch (err: OutOfMemoryError) {
                if (retry == 1) {
                    LOG.d { "Retry reallocate call for OOM" }
                    System.gc()
                } else {
                    throw IOException(err)
                }
            }
            ++retry
        }
    }

    private fun allocateDirectBuffer(size: Long): ByteBuffer {
        validateSize(size)
        var retry = 1
        while (true) {
            try {
                return ByteBuffer.allocateDirect(size.toInt())
            } catch (err: OutOfMemoryError) {
                if (retry == 1) {
                    LOG.d { "Retry reallocate call for OOM" }
                    System.gc()
                } else {
                    throw IOException(err)
                }
            }
            ++retry
        }
    }

    @JvmStatic
    fun copy(source: InputStream, out: OutputStream): Long = source.use { src ->
        val buffer = reallocate(null)
        var total = 0L
        while (true) {
            val size = readPartialContent(src, buffer, 0)
            out.write(buffer, 0, size)
            total += size.toLong()
            if (size != buffer.size) {
                break
            }
        }
        return total
    }

    @JvmStatic
    fun touch(file: File) {
        if (!file.setLastModified(System.currentTimeMillis())) {
            touchFallback(file)
        }
    }

    private fun touchFallback(file: File) = try {
        RandomAccessFile(file, "rw").use {
            val size = it.length()
            it.setLength(size + 1)
            it.setLength(size)
        }
    } catch (e: IOException) {
        LOG.w(e) { "Failed to update file timestamp" }
    }

    @JvmStatic
    fun rename(oldName: File, newName: File) = if (Build.VERSION.SDK_INT >= 26) {
        renameViaFiles(oldName, newName)
    } else {
        (oldName.renameTo(newName)
                || newName.exists() && newName.delete() && oldName.renameTo(newName))
    }

    @RequiresApi(26)
    private fun renameViaFiles(oldName: File, newName: File) = try {
        Files.move(
            oldName.toPath(),
            newName.toPath(),
            StandardCopyOption.ATOMIC_MOVE,
            StandardCopyOption.REPLACE_EXISTING
        )
        true
    } catch (e: Exception) {
        LOG.w(e) { "Failed to rename '${oldName.absolutePath}' -> '${newName.absolutePath}'" }
        false
    }

    @JvmStatic
    fun createByteBufferInputStream(buf: ByteBuffer): InputStream = object : InputStream() {
        override fun available() = buf.remaining()

        override fun mark(readLimit: Int) {
            buf.mark()
        }

        override fun reset() {
            buf.reset()
        }

        override fun markSupported() = true

        override fun read() = if (buf.hasRemaining()) buf.get().toInt() else -1

        override fun read(bytes: ByteArray, off: Int, len: Int) = if (buf.hasRemaining()) {
            val toRead = minOf(len, buf.remaining())
            buf[bytes, off, toRead]
            toRead
        } else {
            -1
        }
    }
}
